### PR TITLE
check if cluster secretRef is nil before using it

### LINKED ISSUE #1621: controller manager crashes when cluster secretRef is empty
**What happened**:

When cluster secretRef is nil, controller manager crashes. 

**What you expected to happen**:

Controller manager should not crash at any time.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a cluster object with secretRef is nil
2. Controller manager crashes

**Anything else we need to know?**:

**Environment**:
- Karmada version:
1.1.1
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
kubectl karmada version: version.Info{GitVersion:"v1.1.1", GitCommit:"8ea4c4d8819dc4c4b5fa885e3b24bb2fef910dbb", GitTreeState:"clean", BuildDate:"2022-03-18T01:34:59Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
- Others:

### PR DESCRIPTION
**Pull request by stingshen**
**What type of PR is this?**

**What this PR does / why we need it**:

Do nil check before using cluster secretRef

**Which issue(s) this PR fixes**:
Fixes #1621 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

```release-note

```

### CODE REVIEW

File: pkg/util/membercluster_client.go [hunk 1]
```diff
@@ -121,12 +121,13 @@ func buildClusterConfig(clusterName string, client client.Client) (*rest.Config,
 		return nil, fmt.Errorf("the api endpoint of cluster %s is empty", clusterName)
 	}
 
-	secretNamespace := cluster.Spec.SecretRef.Namespace
-	secretName := cluster.Spec.SecretRef.Name
-	if secretName == "" {
+	if cluster.Spec.SecretRef == nil || cluster.Spec.SecretRef.Name == "" {
```
Comment (RainbowMango) [inline pkg/util/membercluster_client.go:124]: Do you think we should check the `Namespace` here?

**Comment by RainbowMango** [comment]
Echo your use case from https://github.com/karmada-io/karmada/issues/1621#issuecomment-1101943380 here:
> Users call the api server to register cluster. The api server creates a cluster object without secretRef, sa, impersonate sa, and doesn't wait for the sa secret to be generated. The controller manager watches for the cluster object and fill in the sa secret if available.

This patch prevents panic from nil reference. But before your `controller manager` fills the `secret` the `cluster status controller` might keep failing as it can't build client
https://github.com/karmada-io/karmada/blob/082c3e182d1a8aa19b59b0d23ec672e2aba6471f/pkg/controllers/status/cluster_status_controller.go#L127-L132

Have you ever found this issue? Can you live with that? I'm thinking about how to deal with it in an elegant way.

File: pkg/util/membercluster_client.go (hunk 1 above)
Comment (stingshen) [reply pkg/util/membercluster_client.go]: Actually, `Name` and `Namespace` fields are required in `SecretRef` schema. So, maybe we don't need check `Name` nor `Namespace` here?

**Comment by stingshen** [comment]
> This patch prevents panic from nil reference. But before your `controller manager` fills the `secret` the `cluster status controller` might keep failing as it can't build client
> 
> https://github.com/karmada-io/karmada/blob/082c3e182d1a8aa19b59b0d23ec672e2aba6471f/pkg/controllers/status/cluster_status_controller.go#L127-L132
> 
> Have you ever found this issue? Can you live with that? I'm thinking about how to deal with it in an elegant way.

Yes, I found this issue and can live with that. I think it is the expected behavior that controller throws error and retries in such circumstances...

File: pkg/util/membercluster_client.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/util/membercluster_client.go]: Oh, you are right.

**Comment by RainbowMango** [comment]
> Yes, I found this issue and can live with that. I think it is the expected behavior that controller throws error and retries in such circumstances...

Hold on for a while. Let me see if there is an elegant way. 

By the way, are you using Karmada in production?

**Comment by stingshen** [comment]
> By the way, are you using Karmada in production?

No. We are just doing research and creating a POC of multi-cluster.

**Comment by RainbowMango** [comment]
I guess we can ignore the Cluster that `SecretRef` hasn't present, like:
```diff
diff --git a/cmd/controller-manager/app/controllermanager.go b/cmd/controller-manager/app/controllermanager.go
index e0653716..bf43bf45 100644                                                                               
--- a/cmd/controller-manager/app/controllermanager.go                                                         
+++ b/cmd/controller-manager/app/controllermanager.go                                                         
@@ -191,14 +191,29 @@ func startClusterStatusController(ctx controllerscontext.Context) (enabled bool,        
        clusterPredicateFunc := predicate.Funcs{                                                              
                CreateFunc: func(createEvent event.CreateEvent) bool {
                        obj := createEvent.Object.(*clusterv1alpha1.Cluster)
+
+                       if obj.Spec.SecretRef == nil {
+                               return false
+                       }
+
                        return obj.Spec.SyncMode == clusterv1alpha1.Push
                },
                UpdateFunc: func(updateEvent event.UpdateEvent) bool {
                        obj := updateEvent.ObjectNew.(*clusterv1alpha1.Cluster)
+
+                       if obj.Spec.SecretRef == nil {
+                               return false
+                       }
+
                        return obj.Spec.SyncMode == clusterv1alpha1.Push
                },
                DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
                        obj := deleteEvent.Object.(*clusterv1alpha1.Cluster)
+
+                       if obj.Spec.SecretRef == nil {
+                               return false
+                       }
+
                        return obj.Spec.SyncMode == clusterv1alpha1.Push
                },
                GenericFunc: func(genericEvent event.GenericEvent) bool {
```

How do you think?

**Comment by stingshen** [comment]
> I guess we can ignore the Cluster that `SecretRef` hasn't present, like:
> 
> ```diff
> diff --git a/cmd/controller-manager/app/controllermanager.go b/cmd/controller-manager/app/controllermanager.go
> index e0653716..bf43bf45 100644                                                                               
> --- a/cmd/controller-manager/app/controllermanager.go                                                         
> +++ b/cmd/controller-manager/app/controllermanager.go                                                         
> @@ -191,14 +191,29 @@ func startClusterStatusController(ctx controllerscontext.Context) (enabled bool,        
>         clusterPredicateFunc := predicate.Funcs{                                                              
>                 CreateFunc: func(createEvent event.CreateEvent) bool {
>                         obj := createEvent.Object.(*clusterv1alpha1.Cluster)
> +
> +                       if obj.Spec.SecretRef == nil {
> +                               return false
> +                       }
> +
>                         return obj.Spec.SyncMode == clusterv1alpha1.Push
>                 },
>                 UpdateFunc: func(updateEvent event.UpdateEvent) bool {
>                         obj := updateEvent.ObjectNew.(*clusterv1alpha1.Cluster)
> +
> +                       if obj.Spec.SecretRef == nil {
> +                               return false
> +                       }
> +
>                         return obj.Spec.SyncMode == clusterv1alpha1.Push
>                 },
>                 DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
>                         obj := deleteEvent.Object.(*clusterv1alpha1.Cluster)
> +
> +                       if obj.Spec.SecretRef == nil {
> +                               return false
> +                       }
> +
>                         return obj.Spec.SyncMode == clusterv1alpha1.Push
>                 },
>                 GenericFunc: func(genericEvent event.GenericEvent) bool {
> ```
> 
> How do you think?

There may be a corner case that cluster object cannot be deleted? The following is the procedure:

1. Create a cluster object with SecretRef
2. Cluster controller adds a `finalizer` to the cluster
3. Somehow, delete the SecretRef
4. Delete the cluster object
5. The cluster cannot be deleted because no controller deletes the `finalizer`

I can live with the behavior, but I don't know whether Karmada community can accept that.

**Comment by RainbowMango** [comment]
@stingshen I'm OK with this patch. Just want to confirm if we can eliminate the surprising error logs.

> The cluster cannot be deleted because no controller deletes the finalizer

Why do you think no controller deletes the finalizer?  The [cluster controller](https://github.com/karmada-io/karmada/blob/5b3ef7d93f63daac0b396ef50fdacb4ab0f5ca48/pkg/controllers/cluster/cluster_controller.go#L81-L83) will remove the finalizer(`karmada.io/cluster-controller`). 

The above patch(https://github.com/karmada-io/karmada/pull/1622#issuecomment-1102170751) adds a filter to `cluster status` controller, it's kind of a short path.

**Comment by stingshen** [comment]
> Why do you think no controller deletes the finalizer? The [cluster controller](https://github.com/karmada-io/karmada/blob/5b3ef7d93f63daac0b396ef50fdacb4ab0f5ca48/pkg/controllers/cluster/cluster_controller.go#L81-L83) will remove the finalizer(`karmada.io/cluster-controller`).

Oh, this patch is fine. The cluster can be deleted correctly on the scenario I just provided. I mistakenly thought the filter is on cluster controller:)

I tested the patch on my enviornment. It works fine except that the error still occurs in the following circumstances:

1. Register a cluster with secretRef. Cluster status controller starts updating cluster status
2. Delete the secretRef from cluster. The cluster status controller still reconciles the cluster because the previous reconcile result is: `controllerruntime.Result{Requeue: true}`
3. The reconcile fails with `cluster xxx does not have a secret name` periodically

I think this is an extremely corner case and the behavior is acceptable. What do you think?

**Comment by RainbowMango** [comment]
/lgtm

> Delete the secretRef from cluster. The cluster status controller still reconciles the cluster because the previous reconcile result is: controllerruntime.Result{Requeue: true}

Do you mean delete the `Secret` object or set `.spec.SecretRef` of `Cluster` object to `nil`? That's different.

If you delete the `Secret` object, yes the cluster status controller will continue the reconciliation and fails when it tries to get the secret.

If you set `.spec.SecretRef` of `Cluster` object to `nil`, the cluster `update` event will be ignored from the queue, thus the status controller won't reconcile it (No more error logs). 
But there is a case where, the `update` event is already in the queue, but hasn't been consumed by `cluster status controller`. If you set the `.spec.SecretRef` of the `Cluster` object to `nil`, the controller will reconcile it and continually failing(re-try).

**Comment by RainbowMango** [comment]
This patch is pretty good now. Hold on to confirm the corner case you mentioned above. Thanks, @stingshen in advance.

**Comment by stingshen** [comment]
> /lgtm
> 
> > Delete the secretRef from cluster. The cluster status controller still reconciles the cluster because the previous reconcile result is: controllerruntime.Result{Requeue: true}
> 
> Do you mean delete the `Secret` object or set `.spec.SecretRef` of `Cluster` object to `nil`? That's different.
> 
> If you delete the `Secret` object, yes the cluster status controller will continue the reconciliation and fails when it tries to get the secret.
> 
> If you set `.spec.SecretRef` of `Cluster` object to `nil`, the cluster `update` event will be ignored from the queue, thus the status controller won't reconcile it (No more error logs). But there is a case where, the `update` event is already in the queue, but hasn't been consumed by `cluster status controller`. If you set the `.spec.SecretRef` of the `Cluster` object to `nil`, the controller will reconcile it and continually failing(re-try).

I mean set `.spec.SecretRef` of `Cluster` object to `nil`. From my observation, the controller reconciles the cluster and continually failing in all circumstances following the procedure I provided. No matter whether the reconcile succeeds, the cluster status controller always requeue the reconcile after ClusterStatusUpdateFrequency.

**Comment by RainbowMango** [comment]
> No matter whether the reconcile succeeds, the cluster status controller always requeue the reconcile after ClusterStatusUpdateFrequency.

Yeah. right.

**Review by RainbowMango — APPROVED** [review]
/approve

Looking forward to hearing your feedback. You can DM me if you need any help. Thanks.

**Pull request merged**